### PR TITLE
Adding extended custom post types library.

### LINKED ISSUE #17: Add extended custom post types
https://github.com/johnbillion/extended-cpts

Just as a basic value add to the CMS, it's a really smart and useful way to add post types and customise admin menus.

Will need some documentation for the basic usage, can be copy pasted to some extent from @johnbillion's repo.

### PR DESCRIPTION
**Pull request by kevinlangleyjr**
Fixes #17

### CODE REVIEW

File: inc/namespace.php [hunk 1]
```diff
@@ -128,6 +128,7 @@ function add_login_logo() {
  */
 function load_plugins() {
 	require_once ROOT_DIR . '/vendor/stuttter/wp-user-signups/wp-user-signups.php';
+	require_once ROOT_DIR . '/vendor/johnbillion/extended-cpts/extended-cpts.php';
```
Comment (roborourke) [inline inc/namespace.php]: I had a hunch and checked the `extended-cpts` composer file and it already autoloads this file: https://github.com/johnbillion/extended-cpts/blob/develop/composer.json#L57

The addon has no side effects so it's fine for it to be loaded that way - basically you can just remove this line!

Comment (joehoyle) [reply inc/namespace.php]: Does `extended-cpts.php` have side effects? If so, I think that might be the only package we have that would automaticlly execute side effects.

Comment (kevinlangleyjr) [reply inc/namespace.php]: @joehoyle there are actually no side effects from `extended-cpts` so I think we'll be good removing that line as @roborourke said. 😄

**Comment by kevinlangleyjr** [comment]
@roborourke I've added a new markdown file in the `docs` directory to document the basic functionality and usage of Extended CPTs. I also referred back to the wiki for the library itself for more information.

Let me know what you think about this now that I've got that added 🙂

File: docs/extended-cpts.md [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+For the full documentation, refer to the [Extended CPTs wiki](https://github.com/johnbillion/extended-cpts/wiki).
+
+## Registering Post Types
+
+The `register_extended_post_type()` function is the core of Extended CPTs. It's a wrapper for WordPress' own `register_post_type()` function, which means that any parameters which are accepted by `register_post_type()` are accepted.
+
+The function's signature looks like this:
+
+```php
+function register_extended_post_type( string $post_type, array $args = [], array $names = [] ) : Extended_CPT
```
Comment (roborourke) [inline docs/extended-cpts.md]: Note to self: need to standardise how we do these function signatures and document that somewhere.

Think other examples like this just omit `function` from the start eg

```suggestion
register_extended_post_type( string $post_type, array $args = [], array $names = [] ) : Extended_CPT
```

File: docs/extended-cpts.md [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+```php
+add_action( 'init', function() {
+
+	register_extended_post_type( 'story', [
+		# Add the post type to the site's main RSS feed:
+		'show_in_feed' => true,
+
+		# Show all posts on the post type archive:
+		'archive' => [
+			'nopaging' => true
```
Comment (roborourke) [inline docs/extended-cpts.md]: Not a huge deal here but it's good if copy & pastable code examples adhere to our coding standards, just one less thing for devs to worry about.
```suggestion
			'nopaging' => true,
```

File: docs/extended-cpts.md [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+		],
+
+		# Add some custom columns to the admin screen:
+		'admin_cols' => [
+			'featured_image' => [
+				'title'          => 'Illustration',
+				'featured_image' => 'thumbnail'
```
Comment (roborourke) [inline docs/extended-cpts.md]:
```suggestion
				'featured_image' => 'thumbnail',
```

File: docs/extended-cpts.md [hunk 5]
```diff
@@ -0,0 +1,312 @@
...
+			],
+			'published' => [
+				'title'       => 'Published',
+				'meta_key'    => 'published_date',
+				'date_format' => 'd/m/Y'
```
Comment (roborourke) [inline docs/extended-cpts.md]:
```suggestion
				'date_format' => 'd/m/Y',
```

File: docs/extended-cpts.md [hunk 6]
```diff
@@ -0,0 +1,312 @@
...
+			],
+			'genre' => [
+				'taxonomy' => 'genre'
```
Comment (roborourke) [inline docs/extended-cpts.md]:
```suggestion
				'taxonomy' => 'genre',
```

File: docs/extended-cpts.md [hunk 7]
```diff
@@ -0,0 +1,312 @@
...
+			]
+		],
+
+		# Add a dropdown filter to the admin screen:
+		'admin_filters' => [
+			'genre' => [
+				'taxonomy' => 'genre'
```
Comment (roborourke) [inline docs/extended-cpts.md]:
```suggestion
				'taxonomy' => 'genre',
```

File: docs/extended-cpts.md [hunk 8]
```diff
@@ -0,0 +1,312 @@
...
+			]
+		]
+	],
+	[
+		# Override the base names used for labels:
+		'singular' => 'Story',
+		'plural'   => 'Stories',
+		'slug'     => 'stories'
```
Comment (roborourke) [inline docs/extended-cpts.md]:
```suggestion
		'slug'     => 'stories',
```

File: docs/extended-cpts.md [hunk 9]
```diff
@@ -0,0 +1,312 @@
...
+	] );
+
+} );
+```
+
+And that easily, we have a 'Stories' post type, with correctly generated labels and post updated messages, three custom columns in the admin area (two of which are sortable), stories added to the main RSS feed, and all stories displayed on the post type archive.
+
+## Registering taxonomies
+
+The `register_extended_taxonomy()` is a wrapper for WordPress' own `register_taxonomy()` function, which means that any parameters which are accepted by `register_taxonomy`() are accepted.
```
Comment (roborourke) [inline docs/extended-cpts.md]:
```suggestion
The `register_extended_taxonomy()` is a wrapper for WordPress' own `register_taxonomy()` function, which means that any parameters accepted by `register_taxonomy()` are accepted.
```

File: docs/extended-cpts.md [hunk 10]
```diff
@@ -0,0 +1,312 @@
...
+
+The function's signature looks like this:
+
+```php
+function register_extended_taxonomy( string $taxonomy, $object_type, array $args = [], array $names = [] ) : Extended_Taxonomy
```
Comment (roborourke) [inline docs/extended-cpts.md]:
```suggestion
register_extended_taxonomy( string $taxonomy, $object_type, array $args = [], array $names = [] ) : Extended_Taxonomy
```

Might be good to note that `$object_type` can be a `string|array` too?

File: docs/extended-cpts.md [hunk 11]
```diff
@@ -0,0 +1,312 @@
...
+
+	# Show this taxonomy in the 'At a Glance' dashboard widget:
+	'dashboard_glance' => true,
+
+	# Add a custom column to the admin screen:
+	'admin_cols' => [
+		'updated' => [
+			'title'       => 'Updated',
+			'meta_key'    => 'updated_date',
+			'date_format' => 'd/m/Y'
```
Comment (roborourke) [inline docs/extended-cpts.md]:
```suggestion
			'date_format' => 'd/m/Y',
```

File: docs/extended-cpts.md [hunk 12]
```diff
@@ -0,0 +1,312 @@
...
+		],
+	],
+
+], [
+	# Override the base names used for labels:
+	'singular' => 'Story',
+	'plural'   => 'Stories',
+	'slug'     => 'tales'
```
Comment (roborourke) [inline docs/extended-cpts.md]:
```suggestion
	'slug'     => 'tales',
```

File: docs/extended-cpts.md [hunk 13]
```diff
@@ -0,0 +1,312 @@
...
+### Example
+
+```php
+register_extended_post_type( 'article', [
+
+	'admin_cols' => [
+		// A featured image column:
+		'featured_image' => [
+			'title'          => 'Illustration',
+			'featured_image' => 'thumbnail'
```
Comment (roborourke) [inline docs/extended-cpts.md]:
```suggestion
			'featured_image' => 'thumbnail',
```

File: docs/extended-cpts.md [hunk 14]
```diff
@@ -0,0 +1,312 @@
...
+		],
+		// The default Title column:
+		'title',
+		// A meta field column:
+		'published' => [
+			'title'       => 'Published',
+			'meta_key'    => 'published_date',
+			'date_format' => 'd/m/Y'
```
Comment (roborourke) [inline docs/extended-cpts.md]:
```suggestion
			'date_format' => 'd/m/Y',
```

File: docs/extended-cpts.md [hunk 15]
```diff
@@ -0,0 +1,312 @@
...
+		],
+		// A taxonomy terms column:
+		'genre' => [
+			'taxonomy' => 'genre'
```
Comment (roborourke) [inline docs/extended-cpts.md]:
```suggestion
			'taxonomy' => 'genre',
```

File: docs/extended-cpts.md [hunk 16]
```diff
@@ -0,0 +1,312 @@
...
+Custom post types registered with WordPress' `register_post_type()` function don't allow for a custom permalink structure, unless you want to dive into the Rewrite API and pull your hair out in the process.
+
+Extended CPTs allows a custom permalink structure to be specified via the `permastruct` parameter in the `rewrite` argument.
+
+### Examples
+
+```php
+register_extended_post_type( 'article', [
+	'rewrite' => [
+		'permastruct' => '/foo/%custom_tax%/%article%'
```
Comment (roborourke) [inline docs/extended-cpts.md]:
```suggestion
		'permastruct' => '/foo/%custom_tax%/%article%',
```

File: docs/extended-cpts.md [hunk 17]
```diff
@@ -0,0 +1,312 @@
...
+	],
+] );
+```
+
+```php
+register_extended_post_type( 'article', [
+	'rewrite' => [
+		'permastruct' => '/articles/%year%/%monthnum%/%article%'
```
Comment (roborourke) [inline docs/extended-cpts.md]:
```suggestion
		'permastruct' => '/articles/%year%/%monthnum%/%article%',
```

File: docs/extended-cpts.md [hunk 18]
```diff
@@ -0,0 +1,312 @@
...
+
+The array keys in the `site_filters` array are used as the names of the query vars, which is `my_foo` and `my_genre` in the example below. Be careful with query var name clashes!
+
+### Example
+
+```php
+register_extended_post_type( 'article', [
+	'site_filters' => [
+		'my_foo' => [
+			'meta_key' => 'foo'
```
Comment (roborourke) [inline docs/extended-cpts.md]:
```suggestion
			'meta_key' => 'foo',
```

File: docs/extended-cpts.md [hunk 19]
```diff
@@ -0,0 +1,312 @@
...
+		],
+		'my_genre' => [
+			'taxonomy' => 'genre'
```
Comment (roborourke) [inline docs/extended-cpts.md]:
```suggestion
			'taxonomy' => 'genre',
```

File: docs/extended-cpts.md [hunk 20]
```diff
@@ -0,0 +1,312 @@
...
+
+The array keys in the `site_sortables` array are used for the `orderby` value, which is `my_foo` and `my_genre` in the example below.
+
+### Example
+
+```php
+register_extended_post_type( 'article', [
+	'site_sortables' => [
+		'my_foo' => [
+			'meta_key' => 'foo'
```
Comment (roborourke) [inline docs/extended-cpts.md]:
```suggestion
			'meta_key' => 'foo',
```

File: docs/extended-cpts.md [hunk 21]
```diff
@@ -0,0 +1,312 @@
...
+		],
+		'my_genre' => [
+			'taxonomy' => 'genre'
```
Comment (roborourke) [inline docs/extended-cpts.md]:
```suggestion
			'taxonomy' => 'genre',
```

File: docs/extended-cpts.md [hunk 22]
```diff
@@ -0,0 +1,312 @@
...
+		'plural'   => 'Stories',
+		'slug'     => 'stories'
+	] );
+
+} );
+```
+
+And that easily, we have a 'Stories' post type, with correctly generated labels and post updated messages, three custom columns in the admin area (two of which are sortable), stories added to the main RSS feed, and all stories displayed on the post type archive.
+
+## Registering taxonomies
```
Comment (roborourke) [inline docs/extended-cpts.md]:
```suggestion
## Registering Taxonomies
```

We've standardised on title casing for the rest of the docs, just gonna add commit suggests to make this one consistent.

File: docs/extended-cpts.md [hunk 23]
```diff
@@ -0,0 +1,312 @@
...
+	'singular' => 'Story',
+	'plural'   => 'Stories',
+	'slug'     => 'tales'
+
+] );
+```
+
+And just like that, we have a 'Stories' taxonomy attached to the Post post type, with correctly generated labels and term updated messages, radio buttons in place of the standard meta box for this taxonomy on the post editing screen, a custom column in the admin area (you need to handle the term meta population yourself), and a count of the terms in this taxonomy in the 'At a Glance' dashboard widget.
+
+### Default arguments for custom taxonomies.
```
Comment (roborourke) [inline docs/extended-cpts.md]:
```suggestion
### Default Arguments For Custom Taxonomies.
```

File: docs/extended-cpts.md [hunk 24]
```diff
@@ -0,0 +1,312 @@
...
+			'taxonomy' => 'genre',
+		],
+	],
+
+] );
+```
+
+For more information regarding admin filters and the availabile column types, please see the Extended CPTs wiki page on [admin filters](https://github.com/johnbillion/extended-cpts/wiki/Admin-filters).
+
+## Custom permalink structure
```
Comment (roborourke) [inline docs/extended-cpts.md]:
```suggestion
## Custom Permalink Structure
```

File: docs/extended-cpts.md [hunk 25]
```diff
@@ -0,0 +1,312 @@
...
+	'rewrite' => [
+		'permastruct' => '/articles/%year%/%monthnum%/%article%'
+	],
+] );
+```
+
+All of [WordPress' built-in rewrite tags](https://codex.wordpress.org/Using_Permalinks#Structure_Tags) are supported, including dates and custom taxonomies.
+
+
+## Query vars for filtering
```
Comment (roborourke) [inline docs/extended-cpts.md]:
```suggestion
## Query Vars For Filtering
```

File: docs/extended-cpts.md [hunk 26]
```diff
@@ -0,0 +1,312 @@
...
+```php
+new WP_Query( [
+	'post_type' => 'article',
+	'my_foo'    => 'bar',
+] );
+```
+
+For more information regarding query vars for filtering and the availabile filter types, please see the Extended CPTs wiki page on [query vars for filtering](https://github.com/johnbillion/extended-cpts/wiki/Query-vars-for-filtering).
+
+## Query vars for sorting
```
Comment (roborourke) [inline docs/extended-cpts.md]:
```suggestion
## Query Vars For Sorting
```

File: docs/extended-cpts.md [hunk 27]
```diff
@@ -0,0 +1,312 @@
...
+	'post_type' => 'article',
+	'orderby'   => 'my_genre',
+	'order'     => 'DESC',
+] );
+```
+
+
+For more information regarding query vars for sorting and the availabile sort fields, please see the Extended CPTs wiki page on [query vars for sorting](https://github.com/johnbillion/extended-cpts/wiki/Query-vars-for-sorting).
+
+## Available filters and actions
```
Comment (roborourke) [inline docs/extended-cpts.md]:
```suggestion
## Available Filters and Actions
```

**Pull request merged**